### PR TITLE
Tool ExtractEntireSurface

### PR DESCRIPTION
**Pull request by TomFischer**
[Documentation preview](https://deploy-preview-2764--ogs.netlify.com/docs/tools/meshing-submeshes/extract-entire-surface/)

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.3.0)
2. [x] Tests covering your feature were added?
3. [x] Any new feature or behavior change was documented?

### CODE REVIEW

File: Applications/Utils/MeshEdit/ExtractEntireSurface.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
+/**
+ * @brief Extracts the surface from the given mesh.
```
Comment (endJunction) [inline Applications/Utils/MeshEdit/ExtractEntireSurface.cpp]: \file missing...

File: MeshLib/Elements/QuadRule4.cpp [hunk 2]
```diff
@@ -12,12 +12,20 @@
...
 #include <logog/include/logog.hpp>
 
+#include "Line.h"
 #include "MathLib/GeometricBasics.h"
-
 #include "MeshLib/Node.h"
 
 namespace MeshLib {
 
+const unsigned QuadRule4::face_nodes[4][2] =
```
Comment (endJunction) [inline MeshLib/Elements/QuadRule4.cpp]: This is almost repeated from the edge_nodes. Quad's don't have faces...

File: MeshLib/MeshEditing/DuplicateMeshComponents.cpp [hunk 3]
```diff
@@ -19,8 +19,8 @@
 
 namespace MeshLib
 {
-
-std::vector<MeshLib::Node*> copyNodeVector(const std::vector<MeshLib::Node*> &nodes)
```
Comment (endJunction) [inline MeshLib/MeshEditing/DuplicateMeshComponents.cpp]: Push the clang-format changes elsewhere (like master). They are not part of the tooling.

File: MeshLib/MeshSurfaceExtraction.cpp [hunk 4]
```diff
@@ -98,9 +97,9 @@ MeshLib::Mesh* MeshSurfaceExtraction::getMeshSurface(
     }
 
     std::vector<MeshLib::Node*> sfc_nodes;
-    std::vector<std::size_t> node_id_map(subsfc_mesh.getNumberOfNodes());
-    get2DSurfaceNodes(sfc_nodes, subsfc_mesh.getNumberOfNodes(), sfc_elements,
-                      node_id_map);
+    std::vector<std::size_t> node_id_map;
+    std::tie(sfc_nodes, node_id_map) =
```
Comment (endJunction) [inline MeshLib/MeshSurfaceExtraction.cpp]:
```c++
auto [sfc_nodes, node_id_map] = ...
```

File: MeshLib/MeshSurfaceExtraction.cpp [hunk 5]
```diff
@@ -435,4 +394,218 @@ bool MeshSurfaceExtraction::createSfcMeshProperties(
...
+            surface_elements, bulk_mesh.getNumberOfNodes());
+
+    // create new elements using newly created nodes and delete temp-elements
+    std::vector<MeshLib::Element*> new_elements;
+    try
+    {
+        new_elements = MeshSurfaceExtraction::createSfcElementVector(
+            surface_elements, surface_nodes, node_id_map);
+    }
+    catch (std::runtime_error const& err)
```
Comment (endJunction) [inline MeshLib/MeshSurfaceExtraction.cpp]: Who is throwing the `runtime_error`?

File: MeshLib/MeshSurfaceExtraction.cpp [hunk 6]
```diff
@@ -435,4 +394,218 @@ bool MeshSurfaceExtraction::createSfcMeshProperties(
...
+    if (!MeshSurfaceExtraction::createSfcMeshProperties(
+            *surface_mesh, bulk_mesh.getProperties(),
+            nodes_to_bulk_nodes_id_map, element_to_bulk_element_id_map))
+    {
+        ERR("Transferring subsurface properties failed.");
+    }
+    return surface_mesh;
+}
+
+void transmitBulkIDsToMesh(
```
Comment (endJunction) [inline MeshLib/MeshSurfaceExtraction.cpp]:
```suggestion
void addBulkIdPropertyToMesh(
```

File: MeshLib/MeshSurfaceExtraction.h [hunk 7]
```diff
@@ -91,13 +97,21 @@ class MeshSurfaceExtraction
         double angle,
         unsigned mesh_dimension);
 
-    /// Functionality needed for getSurfaceNodes() and getMeshSurface()
-    static void get2DSurfaceNodes(
-        std::vector<MeshLib::Node*>& sfc_nodes,
-        std::size_t n_all_nodes,
-        const std::vector<MeshLib::Element*>& sfc_elements,
-        std::vector<std::size_t>& node_id_map);
+    static std::tuple<std::vector<MeshLib::Node*>, std::vector<std::size_t>>
```
Comment (endJunction) [inline MeshLib/MeshSurfaceExtraction.h]: Can this static member functions be free (functions)?

File: ProcessLib/HT/Tests.cmake [hunk 8]
```diff
@@ -449,9 +449,9 @@ AddTest(
     VIS ThermalConvection_pcs_0_ts_1_t_0.000000.vtu
 )
 
-if(NOT OGS_USE_MPI AND BUILD_TESTING AND Python_FOUND)
+if(NOT OGS_USE_MPI AND BUILD_TESTING AND Python3_FOUND)
```
Comment (endJunction) [inline ProcessLib/HT/Tests.cmake]: Why the change and why in this PR?

File: web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc [hunk 9]
```diff
@@ -0,0 +1,49 @@
++++
+date = "2020-01-17T5:56:57+01:00"
+title = "Extract Entire Surface"
```
Comment (endJunction) [inline web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc]: I suggest to properly name what is now called "surface"---a 2d entitiy---a boundary, which could be 0d to 2d (for given 3d spaces).

File: web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+## Usage
+
+```
+ExtractEntireSurface -i [<file name of input mesh>] [-o <file name of output mesh>]
+    [--face-property-name <string>]
+    [--element-property-name <string>]
+    [--node-property-name <string>]
+```
+
+The data arrays added to the surface mesh by using the options `--face-property-name` (default value 'bulk_face_ids'), `--element-property-name` (default value 'bulk_element_ids'), and `--node-property-name` (default value 'bulk_node_ids') are used in other tools (for instance in [ComputeNodeAreasFromSurfaceMesh]({{< ref "compute-node-areas-from-surface-mesh" >}})) and is required for flux calculations during a simulation run of OpenGeoSys.
```
Comment (endJunction) [inline web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc]: This is a rather long line. Please add few breaks.
```suggestion
The data arrays added to the surface mesh by using the options `--face-property-name` (default value 'bulk_face_ids'), `--element-property-name` (default value 'bulk_element_ids'), and `--node-property-name` (default value 'bulk_node_ids') are used in other tools (for instance in [ComputeNodeAreasFromSurfaceMesh]({{< ref "compute-node-areas-from-surface-mesh" >}})) and is required for flux calculations during a simulation run of (this is difficult teo review otherwise) OpenGeoSys.
```

**Comment by endJunction** [comment]
...especially in the tool documentation toc it reads:
 - Extract Entire Surface
 - Extract Surface

suggesting, that the second tool does extract only some parts of a surface.

File: MeshLib/Elements/QuadRule4.cpp (hunk 2 above)
Comment (TomFischer) [reply MeshLib/Elements/QuadRule4.cpp]: Changed implementation.

File: MeshLib/MeshSurfaceExtraction.h (hunk 7 above)
Comment (TomFischer) [reply MeshLib/MeshSurfaceExtraction.h]: Yes, most (maybe all) of the static member functions can be free. I don't want to break the 'usual' behavior in this file. I even think that class is unnecessary.

File: ProcessLib/HT/Tests.cmake (hunk 8 above)
Comment (TomFischer) [reply ProcessLib/HT/Tests.cmake]: I will extract it.

File: web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc (hunk 9 above)
Comment (TomFischer) [reply web/content/docs/tools/meshing-submeshes/extract-entire-surface/index.pandoc]: Will rename the tool and change it in the docu.

File: MeshLib/MeshSurfaceExtraction.cpp (hunk 5 above)
Comment (TomFischer) [reply MeshLib/MeshSurfaceExtraction.cpp]: The function createSfcElementVector calls OGS_FATAL that can under certain circumstances throw a `runtime_error`.

**Comment by endJunction** [comment]
Looks better! Few new warnings appeared though.

**Comment by TomFischer** [comment]
Yes, I saw the warnings. Do you have any idea how to fix them?

**Comment by TomFischer** [comment]
Fixed the warnings.

**Comment by endJunction** [comment]
Needs discussion. How is the new method different from the `MeshLib::MeshSurfaceExtraction::getMeshBoundary(mesh)`?

**Comment by TomFischer** [comment]
> Needs discussion. How is the new method different from the `MeshLib::MeshSurfaceExtraction::getMeshBoundary(mesh)`?

 - The new method treats 2D and 3D in the same manner.
 - The new method computes also the `face_id` which is necessary for 2D flux output.
 - The new method is is something like the combination of `MeshLib::MeshSurfaceExtraction::getMeshBoundary(mesh)` and `MeshLib::MeshSurfaceExtraction::getMeshSurface(mesh)`

**Comment by TomFischer** [comment]
Some import new features implemented in the algorithm is:
 - multi-component properties are also extracted
 - higher order boundary elements are created from higher order bulk elements

**Comment by endJunction** [comment]
Other reviews?
Looks good. I especially like that there is a single place for handling the element types. :+1:

**Comment by TomFischer** [comment]
@endJunction  Thanks for reviewing. I think the tool is now much more general (support for higher order elements and multi-component properties output).

**Comment by endJunction** [comment]
@TomFischer Will merge tonight, if there is no other review coming.
